Add best practices for import page (#525)
memgraph/documentation: pages/data-migration/best-practices.mdx

**To speed up import, use `IN_MEMORY_ANALYTICAL` storage mode**. 

Data import is faster in `IN_MEMORY_ANALYTICAL` storage mode because `Delta`
objects are disabled there, which ensures there is no memory overhead. The
`Delta` objects track all changes, and that requires a lot of memory in case of
frequent updates. Not having `Deltas` when importing data, speeds up the process
significantly. 

The `Delta` objects are crucial to ensure atomicity, consistency, isolation and
durability. If you are follwing instuctions, your import process does not need to depend on ACID properties, 
and it is safe to use `IN_MEMORY_ANALYTICAL` storage mode, even in concurrent import scenarios.  

Once you’re done with the data import process, if you have write enabled workloads you should consider
switching back to the `IN_MEMORY_TRANSACTIONAL` mode, which favors
strongly-consistent ACID transactions. On a larger scale import, if you stick with the default
`IN_MEMORY_TRANSACTIONAL` storage mode, you’ll experience slower import time and
spend more resources during data import because of `Delta` objects. 

> TODO(@antejavor): Do we need a link to estimates in possible performance numbers? 

Reviewer: ```suggestion
```

Reviewer: I would drop this, not to block the merge today, and linked numbers once we have them.

`LOAD CSV` queries to import the data. 

Batching the CSV file enables you to run multiple `LOAD CSV` commands in
parallel, which utilizes hardware cores. You should consider batching the CSV
files if you have more than 10 million graph objects.

You need to store nodes and relationships in separate CSV
files. The preparation work for batching includes splitting
those files into smaller CSV files. The size of each CSV file (batch) depends on
your hardware specifications and the total amount of data you are trying to
import. 

From the data perspective, if you have 100 milion nodes, it make sense to split the nodes into batches of 10 or 5 milion nodes per batch. 
You want to avoid having too large files, since loading a single file with 100 milion nodes can create significant memory overhead in multiple scenarios.

From a hardware perspective, if you have a 32-core CPU, it makes sense to split
CSV file into 32 or more batches to fully utilize your machine.
If you have fewer files, you won't put all threads to work, and consequently, you
won't achieve the best import speed.  

To use CSV batching, Memgraph needs to be in the `IN_MEMORY_ANALYTICAL` storage

> mode, since [`SerializationError`](#how-to-handle-a-serializationerror) will

Reviewer: Maybe we can change this to conflicts also?

Depending on the storage mode in which Memgraph is running, the following steps
will differ a bit. **The fastest and least resource-intensive way is the in-memory
analytical storage mode. It is also recommended for running an import via
Cypher.** 

If you are running Cypher import in `IN_MEMORY_ANALYTICAL` mode, your function for
executing chunk of nodes or relationships looks like following: 

```python
def process_chunk_managed_API(query, create_list):
    driver = GraphDatabase.driver(HOST_PORT, auth=("", ""))
    with driver.session(max_transaction_retry_time=180.0, initial_retry_delay=0.2, retry_delay_multiplier=1.1, retry_delay_jitter_factor=0.1) as session:
        session.execute_write(lambda tx: tx.run(query, {"batch": create_list}))
    driver.close()
```

Notice the individual batch being passed to the query. 

If you are running Cypher import in the `IN_MEMORY_ANALYTICAL` mode and using
Memgraph Python driver (pymgclient or GQLAlchemy), the execution of the node
queries won’t differ compared to the `process_chunk` described above. On the
other side, if Memgraph is in `IN_MEMORY_TRANSACTIONAL` mode, the execution of
relationship creation queries can run into

> [`SerializationError`](#how-to-handle-a-serializationerror).

Reviewer: Same thing here.

If you have a file with Cypher queries from another graph database, you can
tweak it to fit Memgraph’s Cypher syntax and import it into Memgraph. Here is an
example of how to do that with the Neo4j Cypher file.

> ## How to handle a `SerializationError`

Reviewer: ```suggestion
## How to handle a conflicting transactions
```

If you're importing data concurrently in the `IN_MEMORY_TRANSACTIONAL` storage

> mode, you can expect `SerializationError` to occur. This happens because your

Reviewer: ```suggestion
mode, you can expect `Cannot resolve conflicting transactions` to occur. This happens because your
```

code tries to concurrently update the same graph object which is locked while
being created or updated. This error is expected and it happens because of ACID
guarantees in `IN_MEMORY_TRANSACTIONAL` storage. 

The dataset structure will highly influence the probability of having a

> `SerializationError` during relationships import. Since nodes are separate

Reviewer: ```suggestion
`Cannot resolve conflicting transactions` error during relationships import. Since nodes are separate
```

entities, this error won’t appear during node import. On the other hand, each
relationship connects two nodes, so if you have a supernode (a node connected to
many other nodes in the dataset), parallel relationship writing on the supernode
will cause conflicts. The more connected the node is, the more serial the
performance will be with more conflicts. 

**There are two most common

> approaches to handle `SerializationError`**:

Reviewer: ```suggestion
approaches to handle conflicting transactions:
```

**1. Switch to in-memory analytical storage mode**

> You won't experience `SerializationError` in the `IN_MEMORY_ANALYTICAL` storage

Reviewer: ```suggestion
You won't experience conflicting transaction error in the `IN_MEMORY_ANALYTICAL` storage
```

If you can't switch to `IN_MEMORY_ANALYTICAL` storage mode and can't avoid
concurrent updates to the same graph objects, then you'll have to retry the
transactions which throw the `SerializationError`. 

```python
def process_chunk(query, create_list, max_retries=100, initial_wait_time=0.200, backoff_factor=1.1, jitter=0.1):
    session = GraphDatabase.driver(HOST_PORT, auth=("", "")).session()
    for attempt in range(max_retries):
        try:
            with session.begin_transaction() as tx:
                tx.run(query, {"batch": create_list})
                tx.commit()
                break
        except TransientError as te:
            jitter = random.uniform(0, jitter) * initial_wait_time 
            wait_time = initial_wait_time * (backoff_factor ** attempt) + jitter
            print(f"Commit failed on attempt {attempt+1}. Retrying in {wait_time} seconds...")
            time.sleep(wait_time)
        except Exception as e:
            print(f"Failed to execute transaction: {e}")
            session.close()
            raise e
```

The execute needs to be in the `try-except` block due to possible exceptions

> from `SerializationError`. If that error happens, the transaction run should be

Reviewer: ```suggestion
from conflicting transactions. If that error happens, the transaction run should be
```

### Use in-memory analytical storage mode

Memgraph supports three storage modes: `IN_MEMORY_TRANSACTIONAL`, `IN_MEMORY_ANALYTICAL` and `ON_DISK_TRANSACTIONAL`

> **To speed up import, use in-memory analytical storage mode**. 

Author: @antejavor Is there any specific reason why you decided for `IN_MEMORY_ANALYTICAL` (and similar) above and left in-memory analytical storage below? Do you think it was wrong to say in-memory transactional, in-memory
analytical and on-disk transactional? I like it when it's consistent.

Reviewer: I do not think it was wrong to say, I just prefer this notation since it is pretty important and to me a bit more visible/important in the code format. Changing a storage mode is a big deal.

Author: I still think we should stick to one notation and link to the instructions on how to change the storage mode if necessary (it's already linked at the bottom of the section). If you want to emphasize it, feel free to put it in bold.

Author: I reviewed our current docs a bit, and you're right, `IN_MEMORY_TRANSACTIONAL`, `IN_MEMORY_ANALYTICAL` and `ON_DISK_TRANSACTIONAL` was often used. So, if you prefer that one, let's stick with that :)

Data import is faster in in-memory analytical storage mode because `Delta`
objects are disabled there, which ensures there is no memory overhead. The
`Delta` objects track all changes, and that requires a lot of memory in case of
frequent updates. Not having `Deltas` when importing data, speeds up the process
significantly. 

The `Delta` objects are crucial to ensure atomicity, consistency, isolation and
durability. If you are follwing instuctions, your import process does not need to depend on ACID properties, 

> and it is safe to use in-memory analytical storage mode, even in concurrent import scenarios.  

Author: There are a couple of typos here. Please check all new content with Grammarly before committing.

Reviewer: Sorry, I typically do that on the whole file, in the end, last step before marking a PR ready for review. Find it a bit faster.

> Once you’re done with the data import process, if you have write enabled workloads you should consider

Author: Can you elaborate more on 'write enabled workloads'?

Reviewer: If you do not have writes in your workload, there is no need to be in-memory transactional mode, since ACID mostly deals with write issues.  Tweaked a sentence a bit to be more clear. 

```suggestion
Once you’re done with the data import process, if you have write queries in your workloads,  consider
```

Author: Much better, thanks

switching back to the in-memory transactional mode, which favors
strongly-consistent ACID transactions. On a larger scale import, if you stick with the default
in-memory transactional storage mode, you’ll experience slower import time and
spend more resources during data import because of `Delta` objects. 

> TODO(@antejavor): Do we need a link to estimates in possible performance numbers? 

Author: We can be explicit. So, feel free to link it.

`LOAD CSV` queries to import the data. 

Batching the CSV file enables you to run multiple `LOAD CSV` commands in
parallel, which utilizes hardware cores. You should consider batching the CSV
files if you have more than 10 million graph objects.

You need to store nodes and relationships in separate CSV
files. The preparation work for batching includes splitting
those files into smaller CSV files. The size of each CSV file (batch) depends on
your hardware specifications and the total amount of data you are trying to
import. 

From the data perspective, if you have 100 milion nodes, it make sense to split the nodes into batches of 10 or 5 milion nodes per batch. 

> You want to avoid having too large files, since loading a single file with 100 milion nodes can create significant memory overhead in multiple scenarios.

Author: Again Grammarly pls 🙏

From a hardware perspective, if you have a 32-core CPU, it makes sense to split
CSV file into 32 or more batches to fully utilize your machine.
If you have fewer files, you won't put all threads to work, and consequently, you
won't achieve the best import speed.  

To use CSV batching, Memgraph needs to be in the in-memory analytical storage
mode, since [`SerializationError`](#how-to-handle-a-serializationerror) will
interrupt the process otherwise. If you are running Memgraph in the `IN_MEMORY_TRANSACTIONAL` mode, 
you will be able to utilize just a single core, and the import process will be much slower and it will 
use more memory because of `Delta` object explanied in [section above](#use-in-memory-analytical-storage-mode).

### Example of batched parallel import with `LOAD CSV` in Python

Assuming that the files have been correctly split, and [Memgraph is

> running](/getting-started) in `IN_MEMORY_ANALYTICAL` mode, the import process can start. If you're running

Author: Let's decide on one way of writing storage mode and stick to that.

**The important factor is that first all nodes needs to be imported. After the last 
file with nodes is imported, the relationships can be imported in the same way.**

## When to use CSV file import tool?

In Memgraph Lab, there is an option of effortlessly importing your data using the
[CSV file import tool](/data-visualization/user-manual/csv-file-import). This tool 
allows you to your data with configuring schema, nodes and relationships with only 
a few clicks.

> Users with a smaller dataset, typically files up to 10MB in size or those less 

Author: ```suggestion
Users with a smaller dataset (typically files up to 10MB in size) or those less 
```

familiar with Cypher queries, will benefit the most from leveraging the CSV file import 
tool. However, for optimal speed and performance, it's still recommended to use 
the `LOAD CSV` Cypher clause. Using the `LOAD CSV` Cypher clause eliminates the 
additional steps in uploading files to Memgraph and communicating with the database.

> For the users with the larger dataset, you can still make use of the CSV import tool 
> by configuring your dataset and utilizing the code generated by the tool. Simply 
> copy and run the generated code directly in the Query execution window and enhance your overall database performance.

Author: ```suggestion
Users with a larger dataset can still make use of the CSV import tool 
by configuring their dataset and utilizing the code generated by the tool. Simply 
copy and run the generated code directly in the Query execution window, enhancing your overall database performance.
```
